src/imitation/algorithms/adversarial.py

```diff
@@ -147,65 +160,104 @@ def gen_policy(self) -> BaseRLModel:
+        self.train_disc_step(gen_samples=gen_samples)
+
+  def train_disc_step(self, *,
+                      gen_samples: Optional[rollout.Transitions] = None,
+                      expert_samples: Optional[rollout.Transitions] = None,
+                      ) -> None:
+    """Perform a single discriminator update, optionally using provided samples.
+
+    Args:
+      gen_samples: Transition samples from the generator policy. If not
+        provided, then take `self.disc_batch_size // 2` samples from the
+        generator replay buffer.
+      expert_samples: Transition samples from the expert. If not
+        provided, then take `n_gen` expert samples from the expert
+        dataset, where `n_gen` is the number of samples in `gen_samples`.
     """
     with logger.accumulate_means("disc"):
-      for _ in range(n_steps):
-        fd = self._build_disc_feed_dict(**kwargs)
-        step, _ = self._sess.run([self._global_step, self._disc_train_op],
-                                 feed_dict=fd)
-        if self._init_tensorboard and step % 20 == 0:
-          self._summarize(fd, step)
-
-  def train_gen(self, n_steps=10000):
+      fd = self._build_disc_feed_dict(gen_samples=gen_samples,
+                                      expert_samples=expert_samples)
+      step, _ = self._sess.run([self._global_step, self._disc_train_op],
+                               feed_dict=fd)
+      if self._init_tensorboard and step % 20 == 0:
+        self._summarize(fd, step)
+
+  def train_gen(self, total_timesteps: Optional[int] = None, callback=None):
+    """Trains the generator (via PPO2) to maximize the discriminator loss.
+
+    After the end of training populates the generator replay buffer (used in
+    discriminator training) with `self.disc_batch_size` transitions.
+
+    Args:
+      total_timesteps: The number of transitions to sample from
+        `self.venv_train_norm` during training. By default,
+        `self.gen_batch_size`.
+      callback: Callback argument to the Stable Baselines `RLModel.learn()`
+        method.
+    """
+    if total_timesteps is None:
+      total_timesteps = self.gen_batch_size
+
     with logger.accumulate_means("gen"):
-      self._gen_policy.set_env(self.venv_train_norm)
+      self.gen_policy.set_env(self.venv_train_norm)
       # TODO(adam): learn was not intended to be called for each training batch
       # It should work, but might incur unnecessary overhead: e.g. in PPO2
       # a new Runner instance is created each time. Also a hotspot for errors:
       # algorithms not tested for this use case, may reset state accidentally.
-      self._gen_policy.learn(n_steps, reset_num_timesteps=False)
+      self.gen_policy.learn(total_timesteps=total_timesteps,
+                            reset_num_timesteps=False,
+                            callback=callback)
       self._populate_gen_replay_buffer()
 
+  def train(self, total_timesteps: int) -> None:
```

Comment:
I like this function but is it used anywhere?

Reply:
No, this isn't referenced outside of test_trainer.py